File: python-algo/leetcode/problem_113.py
"""113. Path Sum II
https://leetcode.com/problems/path-sum-ii/
"""
from typing import List
from common.tree_node import TreeNode


class Solution:
    def pathSum(self, root: TreeNode, targetSum: int) -> List[List[int]]:
        def dfs(node: TreeNode, target: int, path: List[int]):
            path.append(node.val)
            if not node.left and not node.right:
                if node.val == target:
                    ans.append(path[:])
            if node.left:
                dfs(node.left, target - node.val, path)
            if node.right:
                dfs(node.right, target - node.val, path)
            path.pop()
        if not root:
            return []
        ans = []
        dfs(root, targetSum, [])
        return ans

File: java-algo/src/main/java/com/leetcode/Problem113.java
package com.leetcode;

import com.common.TreeNode;

import java.util.ArrayList;
import java.util.List;

/**
 * 113. Path Sum II
 * https://leetcode.com/problems/path-sum-ii/
 */
public class Problem113 {
    private List<List<Integer>> ans = new ArrayList<>();

    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        if (root == null) return ans;
        dfs(root, new ArrayList<>(), targetSum);
        return ans;
    }

    private void dfs(TreeNode node, List<Integer> path, int target) {
        path.add(node.val);
        target -= node.val;
        if (node.left == null && node.right == null) {
            if (target == 0) ans.add(new ArrayList<>(path));
        }
        if (node.left != null) dfs(node.left, path, target);
        if (node.right != null) dfs(node.right, path, target);
        path.remove(path.size() - 1);
    }
}
